cubefs/cubefs
# feat(cli): support unit test for cli

## **What this PR does / why we need it**:
Support unit test for cli. Code refactoring of #2279

## How to mock data
I create fake clients and inject them into the sdk to intercept the request and return the mock data. So we can easily mock the data in test function.

```go
fakeClient := fake.CreateHTTPClient(func(req *http.Request) (*http.Response, error) {
	switch m, p := req.Method, req.URL.Path; {

	case m == http.MethodGet && p == "/admin/aclOp":
		return &http.Response{StatusCode: http.StatusOK, Header: defaultHeader(), Body: fake.SuccessJsonBody(successV1)}, nil

	default:
		t.Fatalf("unexpected request: %#v\n%#v", req.URL, req)
		return nil, nil
	}
})
```

## How to record errors

I rewrited the `errout` function to support modifying the handling of errors. By default, when the `errout` function is called, it will automatically execute the `OsExitWithLogFlush` function, and the program will stop immediately. When test files start running, we will change the `erroutHandler` so that program won't stop and the error will be recorded.

```go
var erroutHandler = OsExitWithLogFlush

func errout(format string, a ...interface{}) {
	log.LogErrorf(format, a...)
	_, _ = fmt.Fprintf(os.Stderr, format, a...)
	erroutHandler(format, a)
}

func OsExitWithLogFlush(_ string, _ ...interface{}) {
	log.LogFlush()
	os.Exit(1)
}
```
```go
func setupTestErrorRecorder() *testErrorRecorder {
	recorder := &testErrorRecorder{}
	erroutHandler = func(format string, args ...interface{}) {
		_, _ = fmt.Fprintf(recorder, format, args...)
	}
	return recorder
}
```

## How to write test cases

I defined a TestCase for unified management of test case execution. In most of the test functions, we only need to write the test case slice and then call the `runTestCases` function for execution.

```go
type TestCase struct {
	name      string    // name of test case
	args      []string  // arguments of test command
	expectErr bool      // whether the error is expected
}

func (c *cliTestRunner) runTestCases(t *testing.T, testCases []*TestCase) {
	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			args := append(c.command, tc.args...)
			err := c.testRun(args...)
			if tc.expectErr {
				assert.Error(t, err)
			} else {
				assert.NoError(t, err)
			}
		})
	}
}
```
```go
func TestAclAddCmd(t *testing.T) {
        // cases for test
	testCases := []*TestCase{
		{
			name:      "Valid arguments",
			args:      []string{"testVol", "192.168.0.1"},
			expectErr: false,
		},
		{
			name:      "missing arguments",
			args:      []string{"testVol"},
			expectErr: true,
		},
	}

        // mock http response
	successV1 := &proto.AclRsp{
		OK: true,
		List: []*proto.AclIpInfo{
			{
				Ip:    "192.168.0.1",
				CTime: 1689091200,
			},
		},
	}

        // fake client for http mock
	fakeClient := fake.CreateHTTPClient(func(req *http.Request) (*http.Response, error) {
		switch m, p := req.Method, req.URL.Path; {

		case m == http.MethodGet && p == "/admin/aclOp":
			return &http.Response{StatusCode: http.StatusOK, Header: defaultHeader(), Body: fake.SuccessJsonBody(successV1)}, nil

		default:
			t.Fatalf("unexpected request: %#v\n%#v", req.URL, req)
			return nil, nil
		}
	})

        // run test
	r := newCliTestRunner().setHttpClient(fakeClient).setCommand("acl", "add")
	r.runTestCases(t, testCases)
}
```
